### PR TITLE
Fix ScaleVM to consider resize volume in any type of service offering

### PR DESCRIPTION
**Pull request by harikrishna-patnala**
### Description

This PR fixes the issue #7325 where root disk is not getting considered for resize in some cases of ScaleVM operation.

While Scaling VM if the compute offering has the root disk size change (for any type of compute offering whether it is normal/constrained/unconstrained) CS should do the resize operation.

We already have a check here (https://github.com/apache/cloudstack/blob/2149e82856f4f658dcec5200430a2ac175b07c91/server/src/main/java/com/cloud/vm/UserVmManagerImpl.java#L2102-L2108) whether to skip the disk offering change or not. So here in this PR I've removed the checks if the compute offering is customized or not which I think irrelavant for root disk resize.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@harikrishna-patnala should this be based of the 4.17 branch?

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan test

**Comment by harikrishna-patnala** [comment]
@kiranchavala can you please test/review this ?

**Comment by kiranchavala** [comment]
@harikrishna-patnala Tested the PR, and the  resize of root disk is working fine after enabling the global setting to true

allow.diskoffering.change.during.scale.vm

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by harikrishna-patnala** [comment]
all these failed tests, report "errorText:Service offering is inactive: " but this seems unrelated to my changes.

Anyone else see these failures elsewhere ?

**Comment by shwstppr** [comment]
@harikrishna-patnala this seems an issue with 4.17 branch itself. I was seeing same failures here, https://github.com/apache/cloudstack/pull/7313#issuecomment-1457426009

**Comment by weizhouapache** [comment]
> @harikrishna-patnala this seems an issue with 4.17 branch itself. I was seeing same failures here, [#7313 (comment)](https://github.com/apache/cloudstack/pull/7313#issuecomment-1457426009)

@harikrishna-patnala @shwstppr 
the issue has been fixed in 4.18/main by #7183 and #7190

I think you can ignore the test failures with 4.17

**Comment by harikrishna-patnala** [comment]
thanks @shwstppr and @weizhouapache

**Comment by kiranchavala** [comment]
@harikrishna-patnala 

When a scale vm  operation is performed in vmware 7. Resize of the root disk is failing on Vmare due to the following exception

`com.cloud.utils.exception.CloudRuntimeException: Failed to resize volume operation of volume UUID: [c0e73278-3955-4981-91f6-0021dba430c1] due to - Failed to resize volume of VM [name: i-2-3-VM] due to: [Resize of volume in VM [name: i-2-3-VM] is not supported because Disk device [path: ROOT-3] has Parents: [6000C295-4f69-b62b-5d38-597886f37804].]`.


I have set vmware.create.full.clone to true 

Could be related to #5992

**Review by kiranchavala — APPROVED** [review]
Tested the feature and its working fine, LGTM

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 1]
```diff
@@ -2022,13 +1982,41 @@ private void validateVolumeResizeWithNewDiskOfferingAndLoad(VolumeVO volume, Dis
...
+                (!newDiskOffering.isCustomized() || (newDiskOffering.isCustomized() && Objects.equals(volume.getSize(), newSize << 30)))) {
+            throw new InvalidParameterValueException(String.format("Volume %s is already having disk offering %s", volume, newDiskOffering.getUuid()));
+        }
+
+        if (existingDiskOffering.getDiskSizeStrictness() != newDiskOffering.getDiskSizeStrictness()) {
+            throw new InvalidParameterValueException("Disk offering size strictness does not match with new disk offering.");
+        }
+
+        if (MatchStoragePoolTagsWithDiskOffering.valueIn(volume.getDataCenterId())) {
+            if (!doesNewDiskOfferingHasTagsAsOldDiskOffering(existingDiskOffering, newDiskOffering)) {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2004]: these can be one condition/if-statement

Comment (harikrishna-patnala) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: Done, updated the if statement

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Review by weizhouapache — APPROVED** [review]
code lgtm

it fixes a valid bug

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by shwstppr — APPROVED** [review]
LGTM

**Comment by DaanHoogland** [comment]
@harikrishna-patnala this gives conflicts on merging forward. Can you have a look please?

**Pull request merged**